### PR TITLE
Dashboard scenes: Remove and duplicate panel options

### PR DESCRIPTION
**Pull request by oscarkilhed**
**What is this feature?**

Enables the non working menu items remove and duplicate panel.

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 1]
```diff
@@ -414,6 +415,36 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
     return this._initialState;
   }
 
+  public duplicateLibraryPanel(libraryPanel: LibraryVizPanel) {
```
Comment (ivanortegaalba) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:418]: Probably we should add a test to `DashboardScene.test.tsx` :)

File: public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx [hunk 2]
```diff
@@ -97,7 +98,7 @@ export function panelMenuBehavior(menu: VizPanelMenu) {
       text: t('panel.header-menu.duplicate', `Duplicate`),
       onClick: () => {
         DashboardInteractions.panelMenuItemClicked('duplicate');
-        dashboard.duplicatePanel(panel);
+        libraryPanel ? dashboard.duplicateLibraryPanel(libraryPanel) : dashboard.duplicatePanel(panel);
```
Comment (ivanortegaalba) [inline public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx:101]: Looking at this, what do you think if we only have `dashboard.duplicatePanel(libraryPanel)` in the dashboard scene and this distinction is managed internally?

File: public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx [hunk 3]
```diff
@@ -377,7 +378,7 @@ function createExtensionContext(panel: VizPanel, dashboard: DashboardScene): Plu
   };
 }
 
-export function removePanel(dashboard: DashboardScene, panel: VizPanel, ask: boolean) {
+export function removePanel(dashboard: DashboardScene, panel: VizPanel | LibraryVizPanel, ask: boolean) {
```
Comment (ivanortegaalba) [inline public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx:381]: Same here, we probably can add the library logic inside this function

**Review by ivanortegaalba — COMMENTED** [review]
Missing the tests for the updated scene objects 🙇

File: public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx (hunk 2 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx:101]: Yeah, maybe. I mean - library panels also respect panel repeat options, which currently ain't handled. I suppose with that in place the difference between these two functions will be quite cosmetic

**Review by dprokop — COMMENTED** [review]
Seems to work, ok. The only thing I've noticed is that the keyboard shortcuts do not work against library panels:
- when duplicating a runtime error is thrown
- when removing nothing happens.

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 4]
```diff
@@ -414,6 +415,36 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
...
+    if (!libraryPanel.parent) {
+      return;
+    }
+    const gridItem = libraryPanel.parent;
+
+    if (!(gridItem instanceof SceneGridItem || PanelRepeaterGridItem)) {
+      console.error('Trying to duplicate a panel in a layout that is not SceneGridItem or PanelRepeaterGridItem');
+      return;
+    }
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:427]: this condition is the same as in duplicatePanel, we should combine this method and duplicatePanel to share this initial logic. Would be nice if panel menu could just call dashboard.duplicatePanel. I do not like that the code that calls duplicatePanel needs to care about if it's a library panel or not

**Review by torkelo — COMMENTED** [review]
would like to move the duplictePanel and duplicateLibraryPanel out of DashboardScene, maybe into a file "gridEditActions"  or something? we can have removePanel there as well. 

duplictePanel can check if parent is a LibraryPanel and handle that case (ie hand it of to separate function if needed). but looks like more can be shared between the two (start and end of the function, just the cloning that is different) 


Tests would be nice.

**Comment by mdvictor** [comment]
I think this can be closed now. Removing was merged duplication is in review [here](https://github.com/grafana/grafana/pull/84159)

**Pull request closed without merging**